### ISSUE TITLE
New map keys aren't indexed in case the table has materialized views

### ISSUE THREAD

**Issue by 633kh4ck** [strapdata/elassandra#278]
**Elassandra version**: 6.2.3 (docker)
**Plugins installed**: []
**JVM version** (`java -version`):
```
openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-8u181-b13-2~deb9u1-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)
```

**OS version** (`uname -a`): `Linux 7f7588b616d2 4.14.85-rancher x86_64 GNU/Linux`

**Description of the problem**:
Similar issue #274 (but not the same). We have a keyspace with a table that has a materialized view. Table has a `map` column and we have an index (elasticsearch mapping) on it. When we perform cassandra insert statement with a payload that contains _new_ object (map) key(s) (key(s) that was never presented inside `map` column in an already existing rows):
- Expected behavior: new data are instantly indexed into elastic and we're able to fetch all the documents (including the new keys in a map/object field) from it;
- Real life: we're able to fetch all the documents (from elastic) with their map fields but only those keys that were indexed earlier (before materialized view creation), in other words - **new map keys aren't indexed in such (with materialized view) case**, never.
So, if we'll drop materialized view and add some more records to this table - everything will work as expected: all the rows are indexed (even with the new keys in the map/object field) and we're able to search by them easily.

<details><summary><strong>Steps to reproduce</strong></summary>

 1. Prepare test keyspace and table
```cql
CREATE KEYSPACE IF NOT EXISTS test WITH replication = {'class': 'NetworkTopologyStrategy', 'DC1': '1'} AND durable_writes = true;
CREATE TABLE IF NOT EXISTS test.data (
    id uuid PRIMARY KEY,
    properties map<text,text>,
    status int
);
```
 2. Add some initial data
```cql
INSERT INTO "test"."data" ("id","properties", "status") VALUES (711cb501-ffaa-1112-12ff-934c51fe54f4, {'fieldname': 'fieldvalue'}, 0);
```
 3. Perform mapping
```bash
curl -X PUT \
  http://localhost:9200/testmapping \
  -H 'content-type: application/json' \
  -d '{
	"settings":{
		"keyspace":"test",
		"number_of_shards":1,
		"number_of_replicas":1
	},
	"mappings":{
		"data":{
			"properties":{
				"properties":{
					"cql_collection":"singleton",
					"cql_struct":"map",
					"type":"object"
				}
			}
		}
	}
}'
```
4. Rebuild index
```bash
nodetool rebuild_index test data elastic_data_idx
```
5. Check that data was fully indexed
```bash
curl -X POST \
  http://localhost:9200/testmapping/_search \
  -H 'content-type: application/json' \
  -d '{
	"query": {
		"match_all": {}
	}
}'
```
<details><summary>Response (ok)</summary>

```json
{
    "took": 1,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": 2,
        "max_score": 1,
        "hits": [
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "711cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue"
                    }
                }
            }
        ]
    }
}
```
</details>

6. Add some more data (with new map key `fieldname2`)
```cql
INSERT INTO "test"."data" ("id","properties", "status") VALUES (111cb501-ffaa-1112-12ff-934c51fe54f4, {'fieldname': 'fieldvalue2', 'fieldname2': 'fieldvalue'}, 0);
```
7. Check documents in index
```bash
curl -X POST \
  http://localhost:9200/testmapping/_search \
  -H 'content-type: application/json' \
  -d '{
	"query": {
		"match_all": {}
	}
}'
```
<details><summary>Response (ok)</summary>

```json
{
    "took": 1,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": 2,
        "max_score": 1,
        "hits": [
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "711cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue"
                    }
                }
            },
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "111cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue2",
                        "fieldname2": "fieldvalue"
                    }
                }
            }
        ]
    }
}
```
</details>

8. Create materialized view
```cql
CREATE MATERIALIZED VIEW databystatus AS
       SELECT status, id FROM data
       WHERE status IS NOT NULL AND id IS NOT NULL
       PRIMARY KEY (status, id);
```
9. Add some more data (with new map key `fieldname3`)
```cql
INSERT INTO "test"."data" ("id","properties", "status") VALUES (211cb501-ffaa-1112-12ff-934c51fe54f4, {'fieldname': 'fieldvalue2', 'fieldname2': 'fieldvalue', 'fieldname3': 'fieldvalue3'}, 0);
```
10. Check documents in index
```bash
curl -X POST \
  http://localhost:9200/testmapping/_search \
  -H 'content-type: application/json' \
  -d '{
	"query": {
		"match_all": {}
	}
}'
```
<details><summary>Response (<strong>data inconsistency</strong>, no <code>fieldname3</code> field in the last document)</summary>

```json
{
    "took": 1,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": 3,
        "max_score": 1,
        "hits": [
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "711cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue"
                    }
                }
            },
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "111cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue2",
                        "fieldname2": "fieldvalue"
                    }
                }
            },
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "211cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue2",
                        "fieldname2": "fieldvalue"
                    }
                }
            }
        ]
    }
}
```
</details>

11. Drop materialized view
```cql
DROP MATERIALIZED VIEW databystatus;
```
12. Add some more data
```cql
INSERT INTO "test"."data" ("id","properties", "status") VALUES (311cb501-ffaa-1112-12ff-934c51fe54f4, {'fieldname': 'fieldvalue2', 'fieldname2': 'fieldvalue', 'fieldname3': 'fieldvalue3', 'fieldname4': 'fieldvalue4'}, 0);
```
13. Check documents in index
```bash
curl -X POST \
  http://localhost:9200/testmapping/_search \
  -H 'content-type: application/json' \
  -d '{
	"query": {
		"match_all": {}
	}
}'
```
<details><summary>Response (<strong>data are fully consistent</strong>)</summary>

```json
{
    "took": 1,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": 4,
        "max_score": 1,
        "hits": [
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "711cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue"
                    }
                }
            },
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "111cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue2",
                        "fieldname2": "fieldvalue"
                    }
                }
            },
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "211cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue2",
                        "fieldname2": "fieldvalue",
                        "fieldname3": "fieldvalue3"
                    }
                }
            },
            {
                "_index": "testmapping",
                "_type": "data",
                "_id": "311cb501-ffaa-1112-12ff-934c51fe54f4",
                "_score": 1,
                "_source": {
                    "properties": {
                        "fieldname": "fieldvalue2",
                        "fieldname2": "fieldvalue",
                        "fieldname3": "fieldvalue3",
                        "fieldname4": "fieldvalue4"
                    }
                }
            }
        ]
    }
}
```
</details>
</details>

**Comment by vroyer** [comment]
This should be fixed in the last release 6.2.3.14 including a fix for #274 
Please let me know if that's not the case.
Thanks.

**Comment by 633kh4ck** [comment]
> This should be fixed in the last release 6.2.3.14 including a fix for #274
> Please let me know if that's not the case.
> Thanks.

Well done, bug was fixed (6.2.3.14), thx :thumbsup:
Can be closed.
